### DISCUSSION TITLE
AES: Plugin vs OpenSSL

### DISCUSSION THREAD

**Discussion by amdrsantos** [strongswan/strongswan#1341] [Q&A] [answered]
What are the differences between using the AES from the AES plugin instead of using the AES from OpenSSL?
- Performance?
- Just using the plugin by default because it is not implemented in other libraries?

Thanks & Regards, Alex

**Comment by tobiasbrunner** [comment, marked as answer]
The _aes_ plugin contains a generic, software-only AES implementation that was provided before we had any plugins that wrapped third-party crypto libraries (it also avoids a dependency on any external libraries). On platforms that support hardware acceleration, it's definitely slower than OpenSSL. Not sure how the software performance compares (OpenSSL might use optimized assembler code). Encryption for IKE isn't that performance-critical, though. But if performance is an issue, the _aes-ni_ plugin provides a generic AES (and AES-GCM) implementation that directly uses the AES-NI instruction set (OpenSSL does the same).

**Reply by amdrsantos** [reply]
I am using the following flags to compile:
```
./configure --host="x86_64-linux-gnu" --enable-monolithic --enable-cmd --disable-ikev1 --enable-openssl --disable-stroke --enable-swanctl --disable-sshkey --disable-dnskey \
  --disable-pgp --disable-pkcs12 --disable-pkcs7 --disable-pkcs1 --disable-random --disable-rc2 --disable-pubkey --disable-curve25519 \
  --disable-xcbc --disable-hmac --disable-gmp --disable-cmac --disable-drbg --disable-attr --enable-error-notify --enable-curl\
  --with-systemdsystemunitdir="$1" --disable-warnings --disable-aes
```

When the charon is initiated, it shows the following messges:

> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[DMN] Starting IKE charon daemon (strongSwan 5.9.8, Linux 4.19.81, x86_64)
> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[LIB] plugin 'aes': failed to load - aes_plugin_create not found and no plugin file available
> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[LIB] plugin 'ctr': failed to load - ctr_plugin_create not found and no plugin file available
> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[LIB] providers loaded by OpenSSL: legacy default
> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[KNL] unable to create IPv4 routing table rule
> Oct 14 12:25:03 xmm4-1-1 charon[199014]: 00[KNL] unable to create IPv6 routing table rule

Nevertheless, ciphers are loaded with OpenSSL:

>   AES_CBC[openssl]
>   AES_CTR[openssl]
>   AES_ECB[openssl]
>   AES_CFB[openssl]

Are the plugin 'xxx' failed to load the messages expected?

**Reply by tobiasbrunner** [reply]
They are if you previously had these two plugins enabled. Check for remaining `aes.conf` and `ctr.conf` snippets that are included in your strongswan.conf (e.g. in `/usr/local/etc/strongswan.d/charon`).

**Reply by amdrsantos** [reply]
That was it, thanks!
